### PR TITLE
Logging added for when compaction takes more than the block time range

### LINKED ISSUE #7763: Improve logging to report point of no return during checkpointing due to overload
**What did you do?**

We run two instances of Prometheus. This pod filled its volume. After expanding the volume the process was continually OOM killed after starting and replaying the WAL. All the while the other instance was running fine. I stopped scraping goldpinger and was able to get past that OOM event and see it enter a loop of Head GC and WAL checkpointing. The checkpoints appear to be getting larger each time. The process did eventually get OOM killed after several hours and started the checkpoint cycle over again.

**What did you expect to see?**

Prometheus recover and match the curve of the other instance.

**What did you see instead? Under which circumstances?**

Checkpoints grow in size. Prometheus never settles into normal disk usage.

**Environment**

* System information:

	Linux 5.5.5-200.fc31.x86_64 x86_64

* Prometheus version:

	prometheus, version 2.19.1 (branch: HEAD, revision: eba3fdcbf0d378b66600281903e3aab515732b39)
	build user:       root@62700b3d0ef9
	build date:       20200618-16:35:26
	go version:       go1.14.4

* Prometheus configuration file:
```
global:
  scrape_interval: 15s

  external_labels:
    cluster: barcelona

rule_files:
- "/rules/alert-rule-loader.rules"


# A scrape configuration for running Prometheus on a Kubernetes cluster.
# This uses separate scrape configs for cluster components (i.e. API server, node)
# and services to allow each to use different authentication configs.
#
# Kubernetes labels will be added as Prometheus labels on metrics via the
# `labelmap` relabeling action.

alerting:
  alertmanagers:
  - kubernetes_sd_configs:
      - role: pod
    tls_config:
      ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
    relabel_configs:
    - source_labels: [__meta_kubernetes_pod_label_app]
      regex: alertmanager
      action: keep
    - source_labels: [__meta_kubernetes_namespace]
      regex: monitoring
      action: keep
    - source_labels: [__meta_kubernetes_pod_container_port_number]
      regex: 9093
      action: keep

# Scrape config for cluster components.
scrape_configs:

- job_name: 'etcd'
  scheme: http
  ec2_sd_configs:
  - region: us-west-2
    port: 9379
  relabel_configs:
  - source_labels: [__meta_ec2_tag_Name]
    regex: barcelona-etcd_nodes-asg-us-west-2[a,b,c]
    action: keep
  - source_labels: [__meta_ec2_tag_aws_autoscaling_groupName]
    target_label: aws_autoscaling_group_name
  - source_labels: [__meta_ec2_private_ip]
    target_label: instance
  - source_labels: [__meta_ec2_instance_id]
    target_label: ec2_instance_id

- job_name: 'docker'
  scheme: http
  ec2_sd_configs:
  - region: us-west-2
    port: 9323
  relabel_configs:
  - source_labels: [__meta_ec2_tag_Name]
    regex: barcelona-(.*?)_nodes-asg-us-west-2[a,b,c]
    action: keep
  - source_labels: [__meta_ec2_tag_aws_autoscaling_groupName]
    target_label: aws_autoscaling_group_name
  - source_labels: [__meta_ec2_private_ip]
    target_label: instance
  - source_labels: [__meta_ec2_instance_id]
    target_label: ec2_instance_id

- job_name: 'kubernetes-cluster'

  # Default to scraping over https. If required, just disable this or change to
  # `http`.
  scheme: https

  # This TLS & bearer token file config is used to connect to the actual scrape
  # endpoints for cluster components. This is separate to discovery auth
  # configuration (`in_cluster` below) because discovery & scraping are two
  # separate concerns in Prometheus.
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
    insecure_skip_verify: true
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token

  kubernetes_sd_configs:
  - role: endpoints

  relabel_configs:
  - source_labels: [__meta_kubernetes_namespace, __meta_kubernetes_service_name, __meta_kubernetes_endpoint_port_name]
    action: keep
    regex: default;kubernetes;https

- job_name: 'kubernetes-nodes'

  scheme: https

  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token

  kubernetes_sd_configs:
  - role: node

  # Attach all node labels to the metrics scraped from the components running
  # on that node.
  relabel_configs:
  - source_labels: [job, __meta_kubernetes_node_label_beta_net_cluster_role]
    action: replace
    target_label: job
    regex: (.+);(.+)
    replacement: $1/$2
  - action: labelmap
    regex: __meta_kubernetes_node_label_(.+)
  - target_label: __address__
    replacement: kubernetes.default.svc:443
  - source_labels: [__meta_kubernetes_node_name]
    regex: (.+)
    target_label: __metrics_path__
    replacement: /api/v1/nodes/${1}/proxy/metrics
  - source_labels: [__meta_kubernetes_node_address_InternalIP]
    action: replace
    target_label: node_ip


- job_name: 'kubelet-probes'

  scheme: https

  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token  

  kubernetes_sd_configs:
  - role: node

  # Attach all node labels to the metrics scraped from the components running
  # on that node.
  relabel_configs:
  - source_labels: [job, __meta_kubernetes_node_label_beta_net_cluster_role]
    action: replace
    target_label: job
    regex: (.+);(.+)
    replacement: $1/$2
  - action: labelmap
    regex: __meta_kubernetes_node_label_(.+)
  - target_label: __address__
    replacement: kubernetes.default.svc:443
  - source_labels: [__meta_kubernetes_node_name]
    regex: (.+)
    target_label: __metrics_path__
    replacement: /api/v1/nodes/${1}/proxy/metrics/probes
  - source_labels: [__meta_kubernetes_node_address_InternalIP]
    action: replace
    target_label: node_ip

# Scrape config for Kubelet cAdvisor.
#
# This is required for Kubernetes 1.7.3 and later, where cAdvisor metrics
# (those whose names begin with 'container_') have been removed from the
# Kubelet metrics endpoint.  This job scrapes the cAdvisor endpoint to
# retrieve those metrics.
#
# In Kubernetes 1.7.0-1.7.2, these metrics are only exposed on the cAdvisor
# HTTP endpoint; use "replacement: /api/v1/nodes/${1}:4194/proxy/metrics"
# in that case (and ensure cAdvisor's HTTP server hasn't been disabled with
# the --cadvisor-port=0 Kubelet flag).
#
# This job is not necessary and should be removed in Kubernetes 1.6 and
# earlier versions, or it will cause the metrics to be scraped twice.
- job_name: 'kubernetes-cadvisor'

  # Default to scraping over https. If required, just disable this or change to
  # `http`.
  scheme: https

  # This TLS & bearer token file config is used to connect to the actual scrape
  # endpoints for cluster components. This is separate to discovery auth
  # configuration because discovery & scraping are two separate concerns in
  # Prometheus. The discovery auth config is automatic if Prometheus runs inside
  # the cluster. Otherwise, more config options have to be provided within the
  # <kubernetes_sd_config>.
  tls_config:
    ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
  bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token

  kubernetes_sd_configs:
  - role: node

  relabel_configs:
  - source_labels: [job, __meta_kubernetes_node_label_beta_net_cluster_role]
    action: replace
    target_label: job
    regex: (.+);(.+)
    replacement: $1/$2
  - action: labelmap
    regex: __meta_kubernetes_node_label_(.+)
  - target_label: __address__
    replacement: kubernetes.default.svc:443
  - source_labels: [__meta_kubernetes_node_name]
    regex: (.+)
    target_label: __metrics_path__
    replacement: /api/v1/nodes/${1}/proxy/metrics/cadvisor
  - source_labels: [__meta_kubernetes_node_address_InternalIP]
    action: replace
    target_label: node_ip

# Scrape config for service endpoints.
#
# The relabeling allows the actual service scrape endpoint to be configured
# via the following annotations:
#
# * `prometheus.io/scrape`: Only scrape services that have a value of `true`
# * `prometheus.io/scheme`: If the metrics endpoint is secured then you will need
# to set this to `https` & most likely set the `tls_config` of the scrape config.
# * `prometheus.io/path`: If the metrics path is not `/metrics` override this.
# * `prometheus.io/port`: If the metrics are exposed on a different port to the
# service then set this appropriately.
- job_name: 'services'

  tls_config:
    # need this to scrape pods that use https, unless we add the roots to this
    # container
    insecure_skip_verify: true

  kubernetes_sd_configs:
  - role: endpoints

  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scrape]
    action: keep
    regex: true
  - source_labels: [__meta_kubernetes_namespace, job, __meta_kubernetes_service_name]
    regex: (.+);(.+);(.+)
    action: replace
    replacement: $1/$2/$3
    target_label: job
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_scheme]
    action: replace
    target_label: __scheme__
    regex: (https?)
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_path]
    action: replace
    target_label: __metrics_path__
    regex: (.+)
  - source_labels: [__address__, __meta_kubernetes_service_annotation_prometheus_io_port]
    action: replace
    target_label: __address__
    regex: (.+)(?::\d+);(\d+)
    replacement: $1:$2
  - action: labelmap
    regex: __meta_kubernetes_service_label_(.+)
  - source_labels: [__meta_kubernetes_namespace]
    action: replace
    target_label: kubernetes_namespace
  - source_labels: [__meta_kubernetes_service_name]
    action: replace
    target_label: service_name

# Example scrape config for probing services via the Blackbox Exporter.
#
# The relabeling allows the actual service scrape endpoint to be configured
# via the following annotations:
#
# * `prometheus.io/probe`: Only probe services that have a value of `true`
- job_name: 'kubernetes-services'

  kubernetes_sd_configs:
  - role: service

  metrics_path: /probe
  params:
    module: [http_2xx]

  relabel_configs:
  - source_labels: [__meta_kubernetes_service_annotation_prometheus_io_probe]
    action: keep
    regex: true
  - source_labels: [__address__]
    target_label: __param_target
  - target_label: __address__
    replacement: blackbox
  - source_labels: [__param_target]
    target_label: instance
  - action: labelmap
    regex: __meta_kubernetes_service_label_(.+)
  - source_labels: [__meta_kubernetes_namespace]
    target_label: kubernetes_namespace
  - source_labels: [__meta_kubernetes_service_name]
    target_label: service_name

# Example scrape config for pods
#
# The relabeling allows the actual pod scrape endpoint to be configured via the
# following annotations:
#
# * `prometheus.io/scrape`: Only scrape pods that have a value of `true`
# * `prometheus.io/port`: Scrape the pod on the indicated port instead of the default of `9102`.
- job_name: 'nodes'
  kubernetes_sd_configs:
  - role: pod

  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
    action: keep
    regex: true
  - source_labels: [__meta_kubernetes_pod_container_name]
    action: keep
    regex: node-exporter
  - source_labels: [job, __meta_kubernetes_pod_label_beta_net_cluster_role]
    regex: (.+);(.+)
    action: replace
    replacement: $1/$2
    target_label: job
  - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
    action: replace
    regex: (.+):(?:\d+);(\d+)
    replacement: ${1}:${2}
    target_label: __address__
  - action: labelmap
    regex: __meta_kubernetes_pod_label_(.+)
  - source_labels: [__meta_kubernetes_pod_node_name]
    action: replace
    target_label: instance
  - source_labels: [__meta_kubernetes_namespace]
    action: replace
    target_label: kubernetes_namespace
  - source_labels: [__meta_kubernetes_pod_host_ip]
    action: replace
    target_label: node_ip
  - source_labels: [__meta_kubernetes_pod_container_name]
    action: replace
    target_label: container_name
  - source_labels: [__meta_kubernetes_pod_container_port_number]
    action: replace
    target_label: container_port_number

- job_name: 'pods'
  kubernetes_sd_configs:
  - role: pod

  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_container_name]
    action: drop
    regex: node-exporter
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
    action: keep
    regex: true
  - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
    action: replace
    regex: (.+):(?:\d+);(\d+)
    replacement: ${1}:${2}
    target_label: __address__
  - source_labels: [__meta_kubernetes_namespace, job, __meta_kubernetes_pod_name]
    regex: (.+);(.+);(.+)
    action: replace
    replacement: $1/$2/$3
    target_label: job
  - source_labels: [__meta_kubernetes_namespace, __meta_kubernetes_pod_name]
    regex: (.+);(.+)
    action: replace
    replacement: $1/$2
    target_label: instance
  - action: labelmap
    regex: __meta_kubernetes_pod_label_(.+)
  - source_labels: [__meta_kubernetes_namespace]
    action: replace
    target_label: kubernetes_namespace
  - source_labels: [__meta_kubernetes_pod_container_name]
    action: replace
    target_label: container_name
  - source_labels: [__meta_kubernetes_pod_container_port_number]
    action: replace
    target_label: container_port_number
  - source_labels: [__meta_kubernetes_pod_node_name]
    action: replace
    target_label: instance
  - source_labels: [__meta_kubernetes_pod_host_ip]
    action: replace
    target_label: node_ip
```

* Logs:
```
level=warn ts=2020-08-05T21:33:47.621Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T21:33:47.755Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T21:33:47.770Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T21:44:51.886Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=91
level=warn ts=2020-08-05T21:44:52.042Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=189
level=warn ts=2020-08-05T21:44:52.060Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T21:44:52.142Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T21:44:52.159Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T21:44:52.176Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T21:44:52.829Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T21:44:52.846Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=info ts=2020-08-05T22:23:38.952Z caller=head.go:869 component=tsdb msg="WAL checkpoint complete" first=23619 last=23899 duration=3h4m37.430014501s
level=info ts=2020-08-05T22:27:04.130Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=15.174800694s
level=warn ts=2020-08-05T22:50:58.447Z caller=manager.go:617 component="rule manager" group=sensors-ORhcD msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T23:32:56.096Z caller=manager.go:617 component="rule manager" group=sensors-ORhcD msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-05T23:43:58.522Z caller=manager.go:617 component="rule manager" group=sensors-ORhcD msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-06T00:56:54.979Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=189
level=warn ts=2020-08-06T00:56:54.988Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=189
level=warn ts=2020-08-06T00:56:55.084Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-06T00:56:55.199Z caller=manager.go:617 component="rule manager" group=sensors-ORhcD msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-06T01:08:00.041Z caller=manager.go:617 component="rule manager" group=sensors-ORhcD msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=info ts=2020-08-06T01:35:52.474Z caller=head.go:869 component=tsdb msg="WAL checkpoint complete" first=23900 last=24057 duration=3h8m48.32313179s
level=info ts=2020-08-06T01:39:14.279Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=14.027002429s
level=warn ts=2020-08-06T01:47:47.594Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=189
level=warn ts=2020-08-06T01:47:47.614Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=189
level=warn ts=2020-08-06T01:47:47.632Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-06T01:47:47.652Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-06T01:47:47.671Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-06T01:47:47.691Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-06T01:47:47.710Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
level=warn ts=2020-08-06T01:47:47.731Z caller=manager.go:617 component="rule manager" group=sensors msg="Error on ingesting out-of-order result from rule evaluation" numDropped=3
```

disk usage graph:

Screen Shot 2020-08-06 at 7 54 35 AM

`promtool debug all` output:

[debug.tar.gz](https://github.com/prometheus/prometheus/files/5038155/debug.tar.gz)

`du` of `wal` directory and checkpoints with `ls -lart`:

[prom.du-ls.txt](https://github.com/prometheus/prometheus/files/5038292/prom.du-ls.txt)

Thanks!

### PR DESCRIPTION
**Pull request by ArthurSens**
Fixes #7763

### CODE REVIEW

**Review by codesome — COMMENTED** [review]
Things are a little more tricky now. WAL truncation + checkpoint happens in `Compact()` now and not here. In `Compact()` I would count the number of head compactions that happened there and then divide (time taken by all head compaction + WAL truncation and checkpointing) with number of head compaction and use that for logging (since we only truncate it once for all Head compactions in the loop). Does that look like a good estimation @beorn7 @brian-brazil?

**Comment by brian-brazil** [comment]
I think that'd be over complicating things, this isn't meant to be 100% precise. A single head compaction taking more than 2h is enough to check for.

The log message is also not quite right. The block compaction did finish (we wouldn't know how long it took otherwise), and I don't think it's wise to advise users to blow away good data.

**Comment by codesome** [comment]
> A single head compaction taking more than 2h is enough to check for.

It is mostly the checkpointing that takes time, compacting Head into a block won't take that long.

**Comment by codesome** [comment]
Did you mean the entire head compaction where we would compact multiple blocks out of Head?

**Comment by brian-brazil** [comment]
No, only one block as that's the usual case. For simplicity let's presume that we've not fallen a block behind already.

**Comment by ArthurSens** [comment]
> Things are a little more tricky now. WAL truncation + checkpoint happens in `Compact()` now and not here.

Didn't know that, thanks for the info!

> The log message is also not quite right. The block compaction did finish (we wouldn't know how long it took otherwise), and I don't think it's wise to advise users to blow away good data.

What would be the best advice here? Ask the user for more resources? Delete something more specific?

> I think that'd be over complicating things, this isn't meant to be 100% precise. A single head compaction taking more than 2h is enough to check for.

I'm a little confused with when to log...  
`(all compaction durations + wal truncation duration) > (lastBlockMaxT-firstBlockMinT)`, is that correct?

**Comment by brian-brazil** [comment]
> What would be the best advice here? Ask the user for more resources? Delete something more specific?

We shouldn't be giving any advice like this in a log message. It's a warning, not an operational manual.

**Comment by codesome** [comment]
@ArthurSens  Ignore what I said, we could do https://github.com/prometheus/prometheus/pull/8151#issuecomment-722241917 to keep it simple for now.

> What would be the best advice here? Ask the user for more resources? Delete something more specific?

Let's just say it's falling behind and won't be able to catch up, no need to tell the solution. I believe there will be scenarios where it can and cannot catch-up and mitigations might be different. We could leave best practices and such things in doc maybe.

File: tsdb/db.go [hunk 1]
```diff
@@ -812,6 +812,7 @@ func (db *DB) Compact() (returnErr error) {
 		mint := db.head.MinTime()
 		maxt := rangeForTimestamp(mint, db.head.chunkRange.Load())
 
+		start := time.Now()
```
Comment (codesome) [inline tsdb/db.go:815]: We should count this entire loop and the WAL truncation. So this should be before the loop.

File: tsdb/db.go [hunk 2]
```diff
@@ -824,6 +825,16 @@ func (db *DB) Compact() (returnErr error) {
...
+		compactionDuration := time.Since(start)
+
+		if compactionDuration.Milliseconds() > maxt-mint {
+			level.Warn(db.logger).Log(
+				"msg", "compaction took longer than the block time range, compactions are falling behind and won't be able to catch up.",
+				"duration", compactionDuration.String(),
+				"mint", mint,
+				"maxt", maxt,
+			)
+		}
```
Comment (codesome) [inline tsdb/db.go:837]: This should be after `db.head.truncateWAL`, and let's use `db.head.chunkRange` instead of `maxt-mint`.

**Review by codesome — COMMENTED** [review]
One concern that I have with this is that TSDB is making assumptions about time. But since this is only for logging a warning, I would take that as _it is fine_.

File: tsdb/db.go (hunk 2 above)
Comment (ArthurSens) [reply tsdb/db.go:837]: Wouldn't this be what I said before? Time spent on multiple compactions plus time spent on head Truncation?

Comment (ArthurSens) [reply tsdb/db.go:837]: And does it make sense to log even if `returnErr != nil`? I think there's no need if an error is already being logged, right?

Comment (ArthurSens) [reply tsdb/db.go:837]: Ignore this last comment, I was thinking about a deferred function for this log, but a deferred function would have to wait for db.compactBlocks() to finish, and that's not what we want

**Comment by ArthurSens** [comment]
Hello :wave:, anything else I should be adjusting here?

File: tsdb/db.go [hunk 3]
```diff
@@ -832,6 +833,14 @@ func (db *DB) Compact() (returnErr error) {
 		return errors.Wrap(err, "WAL truncation in Compact")
 	}
 
+	compactionDuration := time.Since(start)
+	if compactionDuration.Milliseconds() > db.head.chunkRange.Load() {
+		level.Warn(db.logger).Log(
+			"msg", "compaction took longer than the chunk time range, compactions are falling behind and won't be able to catch up.",
```
Comment (codesome) [inline tsdb/db.go:839]:
```suggestion
			"msg", "Head compaction took longer than the block time range, compactions are falling behind and won't be able to catch up",
```

File: tsdb/db.go [hunk 4]
```diff
@@ -832,6 +833,14 @@ func (db *DB) Compact() (returnErr error) {
...
+			"duration", compactionDuration.String(),
+			"chunkRange", db.head.chunkRange.Load(),
```
Comment (codesome) [inline tsdb/db.go:841]:
```suggestion
			"block_range", db.head.chunkRange.Load(),
```

**Review by codesome — APPROVED** [review]
LGTM with some nits

**Comment by ArthurSens** [comment]
Nits addressed :)

**Comment by bwplotka** [comment]
> + go-fuzz-build -libfuzzer -func FuzzParseMetric -o promql-parse-metric.a ./promql
/go/src/github.com/prometheus/prometheus/tsdb/db.go:837:24: compactionDuration.Milliseconds undefined (type time.Duration has no field or method Milliseconds)
typechecking of ./promql failed

File: tsdb/db.go [hunk 5]
```diff
@@ -832,6 +833,15 @@ func (db *DB) Compact() (returnErr error) {
 		return errors.Wrap(err, "WAL truncation in Compact")
 	}
 
+	compactionDuration := time.Since(start)
+	// TODO: change to milliseconds once fuzzit tests are removed
```
Comment (bwplotka) [inline tsdb/db.go:837]:
```suggestion
	// TODO: change to milliseconds once fuzzit tests are removed.
```

Comment (bwplotka) [reply tsdb/db.go:837]: We can fix later.

**Review by bwplotka — APPROVED** [review]
LGTM

**Comment by bwplotka** [comment]
Thanks :muscle:

**Pull request merged**